CIPs/CIP-124/cip-124.md

```diff
@@ -0,0 +1,567 @@
+
+This was rejected since it did not allow for a node to follow a limited subset of the streams.
+
+Predicated LibP2P PubSub channels
+
+Can we change LibP2P PubSub to only send the events that a node cares about to limit the number of messages a node needs to process.
+
+If we can use a TTE(Time To Expire) as part of the predicate we may be able to solve the network storm problem.
+
+This was rejected because it does not solve the missed messages problem.
+
+We still plan to revisit this in the future
+
+#### **DHT for interest discovery**
+In order to discover other ceramic nodes that have overlapping interest
+a node could announce itself in the DHT as interested in a model.
+the dht key would be derived by `separatorKey` and `separatorValue`
+```
+key =  are set, use `partition = hash(separatorKey | separatorValue)`
+```
+
+---
+#### **Stream set sort order**
+example lexicographical sort order
+```
+ceramic/<network>/<partition_key>/<EventID>
+```
+The `partition_key` would be defined for a specific stream set.
+
+For the default stream set with all streams we could partition on`stream type`, `model`, `controller`, `time` , or something else using a partition key? We could even map multiple dimensions into a single partition key and then partition based on the calculated partition key.
+![model controller time cube](assets/cube.png)
+
+#### **What are the sorting orders that make sense?**
+- Total time ordering of all anchored events in all streams.
+    - `ceramic/<network>/<timeID>/<EventID>`
+    - advantages: most divergence is in recent events
+    - disadvantages:
+      - hard to shard as most load is recent events,
+      - hard to follow a stream,
+      - hard to follow a controller 
+- Sort by StreamID lexicographic
+    - `ceramic/<network>/<StreamID>/<EventID>`
+    - advantages: simple to shard to StreamID ranges
+    - disadvantages: following a model or controller would be many ranges
+- Sort by model/StreamID
+    - `ceramic/<network>/<model>/<StreamID>/<event hight><Event CID>`
+    - advantages:
+      - following a models is a single range
+    - disadvantages:
+      - following a controllers is a many ranges
+
+- Sort by model_group/model/stream_group/StreamID/time/event
+    - `ceramic/<network>/<model controller>/<model>/<stream controller>/<StreamID>/<event hight>/<Event CID>`
+    - advantages: models grouped by publisher, streams grouped by publisher, stream events in order
+    - disadvantages: long keys
+
+- Z-order (Hilbert Curve order)
+    - `ceramic/<network>/< z(model, controller, time) >`
+    - advantages: 
+      - clustering by model, controller, and time
+    - disadvantages:
+      - complexity
+      - all three dimensions have few ranges but none are single range.
+    - notes:
+      - [https://youtu.be/YLVkITvF6KU](https://youtu.be/YLVkITvF6KU)
+      - [https://en.wikipedia.org/wiki/Z-order](https://en.wikipedia.org/wiki/Z-order)
```

Comment:
wrong link, correct wiki page for Hilbert Curve
https://en.wikipedia.org/wiki/Z-order_(curve)